File: src/main/java/com/brett/lec17/Lec17Main.java
package com.brett.lec17;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Predicate;

public class Lec17Main {
    public static void main(String[] args) {
        /**
         * 코틀린에서의 람다
         */
        List<Fruit> fruits = Arrays.asList(
                new Fruit("사과", 1_000),
                new Fruit("사과", 1_200),
                new Fruit("사과", 1_200),
                new Fruit("사과", 1_500),
                new Fruit("바나나", 3_000),
                new Fruit("바나나", 3_200),
                new Fruit("바나나", 2_500),
                new Fruit("수박", 10_000)
        );

        /**
         * Closure
         */
         String targetFruitName = "바나나";
//         targetFruitName = "수박";
         filterFruits(fruits, (fruit) -> targetFruitName.equals(fruit.getName()));
         // variables referenced from a lambda expression must be final or effectively final 에러 발생
         // final인 변수 혹은 실제로 final인 변수만 사용가능
    }

    private static List<Fruit> filterFruits(List<Fruit> fruits, Predicate<Fruit> fruitFiler){
        List<Fruit> results = new ArrayList<>();
        for(Fruit fruit : fruits){
            if(fruitFiler.test(fruit)){
                results.add(fruit);
            }
        }
        return results;
    }
}

File: src/main/kotlin/com/brett/lec17/Lec17Main.kt
package com.brett.lec17

fun main(){
    val fruits = listOf(
        Fruit("사과", 1_000),
        Fruit("사과", 1_200),
        Fruit("사과", 1_200),
        Fruit("사과", 1_500),
        Fruit("바나나", 3_000),
        Fruit("바나나", 3_200),
        Fruit("바나나", 2_500),
        Fruit("수박", 10_000),
    )

    /*
     * fun 다음에 함수이름이 들어가야하지만 없음 = 람다
     */
    // 람다를 만드는 방법 1
    val isApple = fun(fruit: Fruit): Boolean{
        return fruit.name == "사과"
    }

    // 람다를 만드는 방법 2
    val isApple2 = {fruit: Fruit -> fruit.name=="사과"}

    // 람다를 직접 호출하는 방법 1
    isApple(fruits[0])

    // 람다를 직접 호출하는 방법 2
    isApple.invoke(fruits[0])

    // 람다에 함수 타입 추가 1
    val isApple3: (Fruit) -> Boolean = fun(fruit: Fruit): Boolean{
        return fruit.name == "사과"
    }

    // 람다에 함수 타입 추가 2
    val isApple4: (Fruit) -> Boolean = {fruit: Fruit -> fruit.name=="사과"}

    filterFruits(fruits, isApple)
    filterFruits(fruits, { fruit: Fruit -> fruit.name=="사과" })
    filterFruits(fruits) { fruit: Fruit -> fruit.name=="사과" }
    filterFruits(fruits) { a -> a.name=="사과" }
    filterFruits(fruits) { it.name=="사과" }

    /**
     * Closure
     */
    var targetFruitName = "바나나"
    targetFruitName = "수박"
    filterFruits(fruits) { it.name == targetFruitName }
}


private fun filterFruits(
    fruits: List<Fruit>, filer: (Fruit) -> Boolean
): List<Fruit>{
    val results = mutableListOf<Fruit>()
    for(fruit in fruits){
        if(filer(fruit)){
            results.add(fruit)
        }
    }
    return results
}